File: 김채린/IT기업 및 대기업 계열사 코테보면서 비슷했던 문제집/영단어 암기는 괴로워.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int[] n = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		HashMap<String, Integer> map = new HashMap<>();
		for(int i = 0; i<n[0]; i++){
			String s = br.readLine();
			if (s.length()>=n[1]){
				map.put(s, map.getOrDefault(s,0)+1);
			}
		}
		StringBuilder sb = new StringBuilder();
		sort(map).forEach(it-> sb.append(it).append("\n"));
		System.out.print(sb);
	}

	public static List<String> sort(Map<String, Integer> map) {
		List<String> list = new ArrayList<>(map.keySet());
		list.sort((a,b) -> {
			if (map.get(a).equals(map.get(b))){
				if (a.length()==b.length()){
					return a.compareTo(b);
				}else{
					return b.length()-a.length();
				}
			}else {
				return map.get(b)-map.get(a);
			}
		});
		return list;
	}
}

File: 김채린/추가문제/24일차/영단어 암기는 괴로워.kt
import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val n = br.readLine().split(" ").map { it.toInt() }
    val map = mutableMapOf<String, Int>()
    for (i in 0 until n[0]) {
        val s = br.readLine()
        if (s.length >= n[1]) {
            map[s] = map.getOrDefault(s, 0) + 1
        }
    }
    val sort =
        map.entries.sortedWith { a, b ->
            when {
                a.value == b.value ->
                    when {
                        a.key.length == b.key.length -> a.key.compareTo(b.key)
                        else -> b.key.length - a.key.length
                    }
                else -> {
                    b.value - a.value
                }
            }
        }
    for (i in sort) {
        bw.write("${i.key}\n")
    }
    bw.flush()
    bw.close()
}
